service/matching/version_rule_helpers.go

```diff
@@ -191,49 +193,140 @@ func DeleteAssignmentRule(timestamp *hlc.Clock,
 	idx := req.GetRuleIndex()
 	actualIdx := given2ActualIdx(idx, rules)
 	if actualIdx < 0 {
-		return nil, serviceerror.NewInvalidArgument(fmt.Sprintf("rule index %d is out of bounds for assignment rule list of length %d", idx, len(getActiveRules(rules))))
+		return nil, serviceerror.NewInvalidArgument(fmt.Sprintf(
+			"rule index %d is out of bounds for assignment rule list of length %d", idx, len(getActiveAssignmentRules(rules))))
 	}
 	rule := rules[actualIdx]
 	rule.DeleteTimestamp = timestamp
 	return data, checkAssignmentConditions(data, 0, req.GetForce(), hadUnfiltered)
 }
 
-// checkRedirectConditions returns an error if there is no room for the new Redirect Rule, or if it causes the
-// queue to fail the other requirements
+// checkRedirectConditions returns an error if the new set of redirect rules don't meet the following requirements:
+// - No more rules than dynamicconfig.VersionRedirectRuleLimitPerQueue
+// - The DAG of redirect rules must not contain a cycle
 func checkRedirectConditions(g *persistencepb.VersioningData, maxRRs int) error {
+	rules := g.GetRedirectRules()
+	if maxRRs > 0 && countActiveRR(rules) > maxRRs {
+		return serviceerror.NewFailedPrecondition(fmt.Sprintf("update would exceed number of redirect rules permitted in namespace dynamic config (%v/%v)", len(rules), maxRRs))
+	}
+	if isCyclic(rules) {
+		return serviceerror.NewFailedPrecondition("update would break acyclic requirement")
+	}
 	return nil
 }
 
 func countActiveRR(rules []*persistencepb.RedirectRule) int {
-	return 0
+	cnt := 0
+	for _, rule := range rules {
+		if rule.DeleteTimestamp == nil {
+			cnt++
+		}
+	}
+	return cnt
+}
+
+// isCyclic returns true if there is a cycle in the DAG of redirect rules.
+func isCyclic(rules []*persistencepb.RedirectRule) bool {
+	dag := makeEdgeMap(rules)
+	for node := range dag {
+		visited := make(map[string]bool)
```

Comment:
I think `visited` can be defined out of the loop. once you visit a node, you don't have to come back to it ever.

Reply:
thanks!